Misc: better instructions for envrc, ignore `/build` instead of `build/`

WaffleLapkin opened the pull request:
See commits for more information.

r? @jieyouxu

WaffleLapkin committed:
add instructions for generating `flake.lock` to `envrc-flake`

Previous setup instructions did not work without. (i.e. the envrc would
not do anything, `nix flake show..` would provide unhelpful error)

WaffleLapkin committed:
ignore `/build` instead of `build/` in gitignore

this does two things:
1. allows making `build` a symlink (which is not considered a directory
   by git, thus removal of trailing `/`).
2. removes the need to special case `rustc_mir_build/src/build`
   (leading `/` makes git only ignore the `build` in the root)

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12074715479/job/33673219886) [(plain)](https://github.com/rust-lang/rust/commit/e0d845a77fc382121f81bcdfbb3cae1d9251c341/checks/33673219886/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.843 Building wheels for collected packages: reuse
#13 2.844   Building wheel for reuse (pyproject.toml): started
#13 3.088   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.089   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.089   Stored in directory: /tmp/pip-ephem-wheel-cache-088mgc5o/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.092 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.499 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.500 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.020 Collecting virtualenv
#13 4.020 Collecting virtualenv
#13 4.073   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.259      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 23.4 MB/s eta 0:00:00
#13 4.312 Collecting platformdirs<5,>=3.9.1
#13 4.321   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.341 Collecting distlib<1,>=0.3.7
#13 4.364      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 37.9 MB/s eta 0:00:00
#13 4.400 Collecting filelock<4,>=3.12.2
#13 4.409   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.409   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.491 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.690 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      341952 kB
DirectMap2M:     8046592 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5687 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_mir_build/src/build/expr/as_rvalue.rs:234: 2-line comment block with odd number of backticks
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Nov 28 19:54:47 UTC 2024
  network time: Thu, 28 Nov 2024 19:54:47 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

WaffleLapkin committed:
fix a comment with uneven number of backticks in rustc_mir_build

this is funny though! apparently tidy parsed `.gitignore`, but did not
recognize unignore lines (`!...`), so tidy was ignoring `rustc_mir_build`
this whole time (at least for some lints?).

WaffleLapkin commented:
> this is funny though! apparently tidy parsed `.gitignore`, but did not recognize unignore lines (`!...`), so tidy was ignoring `rustc_mir_build` this whole time (at least for some lints?).

jieyouxu reviewed (approved):
Thanks!

jieyouxu commented:
FYI @pnkfelix since you expressed previously that you wanted to run `./x` from within subdirectories, and this will modify the gitignore behavior such that it might regress such workflow. However, having gitignore need to ignore directories named `build/` but then also using a negated rule to skip the mir `build` module (not properly supported by many tools incl. github ui and tidy) makes this really weird.

Thus, I'm inclined to only ignore `/build` directory from checkout source root and not any further subdirectories.

@bors r+ rollup

bors commented:
:pushpin: Commit 8e7734978245522cbbd14e53e08e888faf031ded has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 8e7734978245522cbbd14e53e08e888faf031ded -->
<!-- homu: {"type":"Approved","sha":"8e7734978245522cbbd14e53e08e888faf031ded","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
